File: algorithms-in-a-nutshell/heapsort.py
import math

def sort(arr):
  buildHeap(arr)
  for i in range(len(arr) - 1, 0, -1):
    temp = arr[0]
    arr[0] = arr[i]
    arr[i] = temp
    heapify(arr, 0, i)
    
  return arr

def buildHeap(arr):
  
  for i in range(len(arr), -1, -1):
    heapify(arr, i, len(arr))

def heapify(arr, index, max):
  largest = index
  left = 2 * index + 1
  right = 2 * index + 2
  print("left: " + str(left) + "\tright: " + str(right) + "\tlargest: " + str(largest))
  print(arr)
  print('heapifying ' + str(index) + ',' + str(max))
  if (left < max and arr[left] > arr[index]):
    largest = left
  if(right < max and arr[right] > arr[largest]):
    largest = right
  if(largest != index):
    temp = arr[index]
    arr[index] = arr[largest]
    arr[largest] = temp
    heapify(arr, largest, max)

def printHeap(arr):
  #because of the shape property the depth is always log2(n)
  depth = math.ceil(math.log2(len(arr)))
  shouldBreak = False
  for i in range(0, depth):
    for j in range(0, 2 ** i):
      index = j + ( 2 ** i) - 1
      if(index >= len(arr)):
        shouldBreak = True
    
    if shouldBreak:
      break

    
    print()

  return

arr = [8, 11, 9, 2, 10, 16]
sort(arr)
print(arr)
printHeap(arr)

File: algorithms-in-a-nutshell/heapsort.c
static void heapify (void **ar, int (*cmp)(const void *, const void *),
                     int idx, int max) {
  int left = 2*idx + 1;
  int right = 2*idx + 2;
  int largest;

  /* Find largest element of A[idx], A[left], and A[right]. */
  if (left < max && cmp (ar[left], ar[idx]) > 0) {
    largest = left;
  } else {
    largest = idx;
  }

  if (right < max && cmp (ar[right], ar[largest]) > 0) {
    largest = right;
  }

  /* If largest is not already the parent then swap and propagate. */
  if (largest != idx) {
    void *tmp;
    tmp = ar[idx];
    ar[idx] = ar[largest];
    ar[largest] = tmp;

    heapify (ar, cmp, largest, max);
   }
}

static void buildHeap (void **ar,
                       int (*cmp)(const void *, const void *), int n) {
  int i;
  for (i = n/2-1; i>=0; i--) {
    heapify (ar, cmp, i, n);
  }
}

void sortPointers (void **ar, int n,
                   int (*cmp)(const void *, const void *)) {
  int i;
  buildHeap (ar, cmp, n);
  for (i = n-1; i >= 1; i--) {
    void *tmp;
    tmp = ar[0];
    ar[0] = ar[i];
    ar[i] = tmp;

    heapify (ar, cmp, 0, i);
  }
}

int[] arr = [8, 11, 9, 2, 10, 16];
sortPointers(arr, 6);